### PR TITLE
Add RPI to GUI packet telecommunication system (real & simulated) to retrieve data from the robot 

### PR DESCRIPTION
**Pull request by sazhang02**
see below

### CODE REVIEW

**Comment by sazhang02** [comment]
Demo of GUI running on simulated packets: [link](https://drive.google.com/file/d/18tQ-ZICiD8dxegr2QiX47QZgor9ABbkS/view?usp=sharing). Video taken robot is asleep for 0.1 for each movement step and packet sender is asleep for 0.01

File: engine/database.py [hunk 1]
```diff
@@ -89,8 +89,19 @@ def phase_as_value(self):
...
+        temp_n_dist = "00.0"
+        temp_rot = "00.00"
+        temp_last_n = ["000.00","000.00"]
+        temp_vel = "0.00"
+        next_n = ["000.00","000.00"]
+        temp_ctrl = "1"
+        packet = Packet(self.phase_as_value(), str(self.get_data("plastic_weight")), acc,\
+               temp_n_dist, temp_rot, temp_last_n, temp_vel, next_n, coords, str(self.get_data("battery")), temp_ctrl)
+
+        return str(packet)
```
Comment (sazhang02) [inline engine/database.py:107]: Creating a packet string using values in the database. Some values are temporary/hardcoded since we do not have them accessible. These should be replaced in the future.

File: engine/mission.py [hunk 2]
```diff
@@ -62,13 +62,14 @@ def __init__(self, robot, base_station, init_control_mode, grid=Grid(42.444250,
...
         self.waypoints_to_visit = deque(self.all_waypoints)
         self.allowed_dist_error = allowed_dist_error
-        self.gps_serial = serial.Serial('/dev/ttyACM0', 19200, timeout=5) 
-        self.robot_radio_serial = serial.Serial('/dev/ttyS0', 57600) #robot radio 
-        self.imu_i2c = busio.I2C(board.SCL, board.SDA)
-        self.motor_controller = MotorController(robot, wheel_r = 0, vm_load1 = 1, vm_load2 = 1, L = 0, R = 0)
-        self.robot_radio_session = RadioSession(self.robot_radio_serial) 
-        self.gps = GPS(self.gps_serial) 
-        self.imu = IMU(self.imu_i2c) 
+        if not robot.is_sim:
```
Comment (sazhang02) [inline engine/mission.py:65]: wrap all rpi dependencies under the is_sim flag

File: engine/packet.py [hunk 3]
```diff
@@ -99,14 +105,14 @@ def __init__(self, phase, p_weight, acc, n_dist, rot, last_n, vel, next_n, coord
...
         args = [("phase:", self.phase), (";p_weight:", fix_data_size(self.p_weight, 2, 1)),
-                (";acc:", (fix_data_size(self.acc[0], 1, 2)+","+fix_data_size(self.acc[1], 1, 2))+","+fix_data_size(self.acc[1], 1, 2)), 
+                (";acc:", fix_values_size(self.acc, 1, 2)), 
                 (";n_dist:", fix_data_size(self.n_dist, 2, 1)),
-                (";rot:", fix_data_size(self.rot, 2, 2)), (";last_n:", fix_tuple_size(self.last_n, 3, 2)),
-                (";vel:", fix_data_size(self.vel, 1, 2)), (";next_n:", fix_tuple_size(self.next_n, 3, 2)),
-                (";coord:", fix_tuple_size(self.coord, 3, 2)),
+                (";rot:", fix_data_size(self.rot, 2, 2)), (";last_n:", fix_values_size(self.last_n, 3, 2)),
+                (";vel:", fix_data_size(self.vel, 1, 2)), (";next_n:", fix_values_size(self.next_n, 3, 2)),
+                (";coord:", fix_values_size(self.coord, 3, 2)),
```
Comment (sazhang02) [inline engine/packet.py:115]: fixing the string version of each value to match the specification of a packet string. For example, the x coord may be 0.0 but to match the specification of the packet string, it should be adjusted to "000.00".

File: gui/gui.py [hunk 4]
```diff
@@ -188,7 +189,7 @@ def setup_gui():
                 [sg.Multiline(current_output, key = "-OUTPUT-", size=(40,8), disabled=True, font=('Courier New', 20))],
                 [sg.Text("Current Coordinates: ______")],
                 [sg.Text("Current Phase: ______", key = "-PHASE-")],
-                [sg.Button('Autonomous', key = "-CONTROL_MODE_BUTTON-"), sg.Button('Track Location'), sg.Button('Traversal Phase'), sg.Button('Simulation')],
+                [sg.Button('Autonomous', key = "-CONTROL_MODE_BUTTON-"), sg.Button('Track Location'), sg.Button('Traversal Phase'), sg.Button('Simulation'), sg.Button('Read RPI Comms')],
```
Comment (sazhang02) [inline gui/gui.py:192]: Read RPI Comms will trigger the gui to start reading any values that come in from the rpi

File: gui/retrieve_inputs.py [hunk 5]
```diff
@@ -1,50 +1,8 @@
 import os
 import sys
-import serial
 import statistics
 from engine.packet import *
-from gui.robot_data import get_tuple_value, get_integer_value, get_float_value
-# ser = serial.Serial("/dev/cu.usbserial-017543DC", 57600)
-
-
-def update_gui():
```
Comment (sazhang02) [inline gui/retrieve_inputs.py:10]: implementation added to gui main loop

File: gui/robot_data.py [hunk 6]
```diff
@@ -34,26 +33,28 @@ def get_float_value(packet):
     Returns: float of "value"
 
     '''
-    return float(get_values(packet))
+    return float(get_value(packet))
 
 
-def get_tuple_value(data):
+def get_values(data, num_inputs):
```
Comment (sazhang02) [inline gui/robot_data.py:39]: make input list rather than tuple so we can store more than 2 values of data (need 3 for acceleration and coordinates)

File: tests/functionality_tests/csv/gui_to_rpi.csv [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+"phase:2;p_weight:00.0;acc:0.00,0.00,0.00;n_dist:00.0;rot:00.00;last_n:000.00,000.00;vel:0.00;next_n:000.00,000.00;coord:000.00,000.00;batt:000;ctrl:1"
+"phase:2;p_weight:00.0;acc:0.00,0.00,0.00;n_dist:00.0;rot:00.00;last_n:000.00,000.00;vel:0.00;next_n:000.00,000.00;coord:000.00,000.00;batt:000;ctrl:1"
+"phase:3;p_weight:01.0;acc:0.00,0.00,0.00;n_dist:00.0;rot:00.00;last_n:000.00,000.00;vel:0.00;next_n:000.00,000.00;coord:000.00,000.00;batt:000;ctrl:1"
+"phase:3;p_weight:01.0;acc:0.00,0.00,0.00;n_dist:00.0;rot:00.00;last_n:000.00,000.00;vel:0.00;next_n:000.00,000.00;coord:000.00,000.00;batt:000;ctrl:1"
+"phase:3;p_weight:01.0;acc:0.00,0.00,0.00;n_dist:00.0;rot:00.00;last_n:000.00,000.00;vel:0.00;next_n:000.00,000.00;coord:000.00,000.00;batt:000;ctrl:1"
+"phase:3;p_weight:01.0;acc:0.00,0.00,0.00;n_dist:00.0;rot:00.00;last_n:000.00,000.00;vel:0.00;next_n:000.00,000.00;coord:000.00,000.00;batt:000;ctrl:1"
+"phase:3;p_weight:01.0;acc:0.00,0.00,0.00;n_dist:00.0;rot:00.00;last_n:000.00,000.00;vel:0.00;next_n:000.00,000.00;coord:000.00,000.00;batt:000;ctrl:1"
```
Comment (sazhang02) [inline tests/functionality_tests/csv/gui_to_rpi.csv:20]: mock data for test case

**Review by sazhang02 — COMMENTED** [review]
comments to explain ambiguous parts of code

**Comment by sazhang02** [comment]
## Summary
### TLDR: Adds a system to communicate between the RPI and GUI for both the real-time case and a simulated case. This will eventually allow us to create an accurate GUI display using the data in our mission as the robot runs.


### Description

In this PR, we established a connection between the RPI and GUI. The RPI can send telemetry packets (string containing data we need as our mission fsm runs on the robot) to the GUI. Then the GUI can retrieve the packets and update it's UI accordingly. Currently, only the location of the robot on the map plot is updated based on the coordinate data sent within the packets.

This connection functions both in a 
1. Real-time environment (Mission FSM on raspberry pi sending data to GUI, communicating using RF modules)
2. Simulated environment (Mission FSM run on the same computer as the GUI at the same time, communicating through a csv file).

There are is_sim flags in place to allow us to switch between these two cases.

## Test Plan

Real-Time environment: [[comment from last spring](https://github.com/cornellnexus/src/pull/72#issuecomment-1119189011)]
- Turned on GUI on laptop, w/ RF module connected.
- In the maker space, ran `gui_to_rpi.py` on the raspberry pi to send a packet to the GUI. 
- Checked that the robot marker on the GUI map moved to the position sent by the packet in `gui_to_rpi.py`
** Note: some changes were made when fixing this code to work for a simulated environment, so this real-time test will need to be done again when the makerspace opens up.

Simulated Environment:
- Turned on GUI on laptop
- Press Simulation button
- Check logs to see that packets are sent, and robot moves along GUI (may happen in a laggy manner due to the speed of internal position calculations, which can be fixed by adding more time to sleep in `robot.py` movement function). It should look like the video in this comment w/ the respective sleep time settings [[comment](https://github.com/cornellnexus/src/pull/72#issuecomment-1238919349)].
## Other

Co-author: @angelaloh09

**Comment by sazhang02** [comment]
@Shreya1233 @AdityaKompella6

**Review by Shreya1233 — COMMENTED** [review]
Everything seemed good! All the necessary changes made sense to me and I didn't see anything else to add.

**Review by AdityaKompella6 — APPROVED** [review]
Sammy you killed it. You are good at renaming stuff.

**Review by Shreya1233 — APPROVED** [review]
